### PR TITLE
Add REST endpoint to get DOT graph of a job

### PR DESCRIPTION
**Pull request by andygrove**
# Which issue does this PR close?

N/A

 # Rationale for this change

This is essential to comprehending how Ballista is executing a query.

test

# What changes are included in this PR?

New endpoint: `http://localhost:50050/api/job/<job-id>/dot`

# Are there any user-facing changes?

### CODE REVIEW

**Review by avantgardnerio — COMMENTED** [review]
Very nice! What generates the png?

**Comment by andygrove** [comment]
> Very nice! What generates the png?

`http://localhost:50050/jobs/<job-id>/dot` serves up the raw dot file. I am converting to png locally. I would like the Web UI to just render this though.

**Comment by avantgardnerio** [comment]
> > Very nice! What generates the png?
> 
> `http://localhost:50050/jobs/<job-id>/dot` serves up the raw dot file. I am converting to png locally. I would like the Web UI to just render this though.

I didn't think this was possible, but it looks like someone has transpiled graphviz to wasm so it runs in the browser now https://github.com/magjac/d3-graphviz.

**Comment by yahoNanJing** [comment]
very nice graph

**Comment by andygrove** [comment]
@yahoNanJing @thinkharderdev @avantgardnerio This is now ready for review. I will follow up with smaller PRs to improve formatting and add metrics.

File: ballista/rust/scheduler/src/state/execution_graph_dot.rs [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+    i: usize,
+) -> Result<StagePlanState, fmt::Error> {
+    let mut state = StagePlanState {
+        readers: HashMap::new(),
+    };
+    write_plan_recursive(f, prefix, plan, i, &mut state)?;
+    Ok(state)
+}
+
+fn write_plan_recursive(
```
Comment (Dandandan) [inline ballista/rust/scheduler/src/state/execution_graph_dot.rs:164]: We could use/move something similar to DataFusion (minus the stage part, currently).

**Comment by yahoNanJing** [comment]
LGTM

**Review by yahoNanJing — APPROVED** [review]

File: ballista/rust/scheduler/src/state/execution_graph_dot.rs [hunk 2]
```diff
@@ -0,0 +1,537 @@
...
+    readers: HashMap<String, usize>,
+}
+
+/// Make strings dot-friendly
+fn sanitize_dot_label(str: &str) -> String {
+    // TODO make max length configurable eventually
+    sanitize(str, Some(100))
+}
+
+/// Make strings dot-friendly
```
Comment (Dandandan) [inline ballista/rust/scheduler/src/state/execution_graph_dot.rs:233]: AFAIK labels can be quoted like this 

`"Thïs įs a 🦄 and a \" quote"`

Comment (Dandandan) [reply ballista/rust/scheduler/src/state/execution_graph_dot.rs:233]: So surround by quotes and escape quote character is all that's needed.

Comment (andygrove) [reply ballista/rust/scheduler/src/state/execution_graph_dot.rs:233]: I did try this, and it failed, but maybe I made a mistake ... I will try again in the next PR

**Comment by andygrove** [comment]
Thanks for the review @Dandandan and @yahoNanJing

**Pull request merged**